tools/genomespace/genomespace_importer.py

```diff
@@ -1,220 +1,265 @@
-            except:
-                pass  # sniff failed
-        if file_type is None and '.' in parsed_url[2]:
-            # still no known datatype, fall back to using extension
-            file_type = parsed_url[2].rsplit( '.', 1 )[-1]
-            file_type = GENOMESPACE_EXT_TO_GALAXY_EXT.get( file_type, file_type )
-        if file_type is None:
-            # use default extension (e.g. 'data')
-            file_type = DEFAULT_GALAXY_EXT
-
-        # save json info for single primary dataset
-        if dataset_id is not None:
+            rval[ key ] = str( value )
+    return rval
+
+
+def exec_before_job( app, inp_data, out_data, param_dict=None, tool=None ):
+    """
+    Galaxy override hook
+    See: https://wiki.galaxyproject.org/Admin/Tools/ToolConfigSyntax#A.3Ccode.3E_tag_set
+    Since only tools with tool_type="data_source" provides functionality for having a JSON param file such as this:
+    https://wiki.galaxyproject.org/Admin/Tools/DataManagers/DataManagerJSONSyntax#Example_JSON_input_to_tool,
+    this hook is used to manually create a similar JSON file.
+    However, this hook does not provide access to GALAXY_DATATYPES_CONF_FILE and GALAXY_ROOT_DIR
+    properties, so these must be passed in as commandline params.
+    """
+    if param_dict is None:
+        param_dict = {}
+    json_params = {}
+    json_params[ 'param_dict' ] = _prepare_json_param_dict( param_dict )
+    json_params[ 'output_data' ] = []
+    json_params[ 'job_config' ] = dict( GALAXY_DATATYPES_CONF_FILE=param_dict.get( 'GALAXY_DATATYPES_CONF_FILE' ), GALAXY_ROOT_DIR=param_dict.get( 'GALAXY_ROOT_DIR' ), TOOL_PROVIDED_JOB_METADATA_FILE=galaxy.jobs.TOOL_PROVIDED_JOB_METADATA_FILE )
+    json_filename = None
+    for i, ( out_name, data ) in enumerate( out_data.iteritems() ):
+        file_name = data.get_file_name()
+        data_dict = dict( out_data_name=out_name,
+                          ext=data.ext,
+                          dataset_id=data.dataset.id,
+                          hda_id=data.id,
+                          file_name=file_name )
+        json_params[ 'output_data' ].append( data_dict )
+        if json_filename is None:
+            json_filename = file_name
+    with open( json_filename, 'w' ) as out:
+        out.write( json.dumps( json_params ) )
+
+
+def get_galaxy_ext_from_genomespace_format(format):
+    return GENOMESPACE_EXT_TO_GALAXY_EXT.get(format, None)
+
+
+def get_galaxy_ext_from_file_ext(filename):
+    if not filename:
+        return None
+    filename = filename.lower()
+    ext = filename.rsplit('.', 1)[-1]
+    return get_galaxy_ext_from_genomespace_format(ext)
+
+
+def sniff_data_type(json_params, output_file):
+    """
+    The sniff.handle_uploaded_dataset_file() method in Galaxy performs dual
+    functions: it sniffs the filetype and if it's a compressed archive like
+    gzip or bz2, decompresses the file, replacing the original file.
```

Comment:
To clarify I think it would be good to change this to say that compressed files for non-compressed datatypes will be unpacked. I think that was part of the confusion.